### ISSUE TITLE
Validity of HTML Lang attribute (bf051a) - incorrect info about xml:lang

### ISSUE THREAD

**Issue by dd8** [act-rules/act-rules.github.io#668]
https://act-rules.github.io/rules/bf051a#accessibility-support

The Accessibility Support section says:

> While HTML5 specification indicates that xml:lang attribute takes priority over lang attribute, certain assistive technologies prioritise lang over xml:lang instead.

The HTML 5.2 and the HTML living standards actually say:

>Authors must not use the lang attribute in the XML namespace on HTML elements in HTML documents. 
>The attribute in no namespace with no prefix and with the literal localname "xml:lang" has no effect on language processing.

https://www.w3.org/TR/html52/dom.html#the-lang-and-xmllang-attributes
https://html.spec.whatwg.org/multipage/dom.html#the-lang-and-xml:lang-attributes

**Comment by Jym77** [comment]
I think you are right in raising the issue, and there are two separate but linked questions.

(Quotes without source are taken from https://www.w3.org/TR/html52/dom.html#the-lang-and-xmllang-attributes)

---

The first issue is the precedence between `lang` and `xml:lang` *when both are valid*. For that, there is a non-ambiguous paragraph:
> If both the `lang` attribute in no namespace and the `lang` attribute in the XML namespace are set on an element, user agents must use the `lang` attribute in the XML namespace, and the `lang` attribute in no namespace must be ignored for the purposes of determining the element’s language.

So `xml:lang` takes priority over `lang`, without any doubt.

---

The second issue is the one you are raising: `xml:lang` should actually not be used in many places…
> The `lang` attribute in the XML namespace may be used on HTML elements in XML documents, as well as elements in other namespaces if the relevant specifications allow it (in particular, MathML and SVG allow `lang` attributes in the XML namespace to be specified on their elements). If both the `lang` attribute in no namespace and the `lang` attribute in the XML namespace are specified on the same element, they must have exactly the same value when compared in an ASCII case-insensitive manner.

> Authors must not use the `lang` attribute in the XML namespace on HTML elements in HTML documents. To ease migration to and from XHTML, authors may specify an attribute in no namespace with no prefix and with the literal localname `xml:lang` on HTML elements in HTML documents, but such attributes must only be specified if a `lang` attribute in no namespace is also specified, and both attributes must have the same value when compared in an ASCII case-insensitive manner.

From that, I get that `<html xml:lang="en">` is not allowed (`xml:lang` without `lang`) and `<html lang="en-us" xml:lang="en-uk">` is not allowed either (mismatching `lang` and `xml:lang`). These are most of our examples in several rules that check language on toplevel of the page:
* [HTML has lang attribute](https://act-rules.github.io/rules/b5c3f8)
* [Validity of HTML Lang attribute](https://act-rules.github.io/rules/bf051a)
* [HTML lang and xml:lang match](https://act-rules.github.io/rules/5b7ae0)

This is, I believe, a lesser issue for [Valid body lang attribute](https://act-rules.github.io/rules/de46e4) which is checking language of specific elements. The examples in that rule are still somewhat wrong (by putting `xml:lang` on HTML elements within an HTML document) but could easily be adapted to, *e.g.* SVG elements, or HTML elements within an XML document, …

---

On a side note, the HTML specification already requires `lang` and `xml:lang` to match exactly:
> If both the lang attribute in no namespace and the lang attribute in the XML namespace are specified on the same element, they must have exactly the same value when compared in an ASCII case-insensitive manner.

which is stronger than what we check in [HTML lang and xml:lang match](https://act-rules.github.io/rules/5b7ae0) where we do have a note mentioning that the rule is less strict than HTML5 specification. Thus we are explicitly testing (and accepting) something that is not valid HTML5. With that regard, accepting `<html xml:lang="en">` might be OK…

We might want to add or correct notes in other rules, *e.g.* [HTML has lang attribute](https://act-rules.github.io/rules/b5c3f8) has a note saying:
> **Note**: HTML5 recommends using `lang` instead of `xml:lang`. This is not known to impact accessibility, which is why use of both is permitted by this rule.

Which is wrong. It’s not "recommended", it's "mandatory".

---

I guess the overall question is: do we want to only test valid HTML5, or do we accept to test invalid HTML5? Which might be rephrased as: do we require developers to fix their HTML5 before turning to accessibility, or do we want to help make more accessible even pages that are invalid HTML5?

**Comment by EmmaJP** [comment]
I've always approached ACT Rules with a notion that it is testing valid HTML, as there are way too many permutations otherwise. 

Our test tool runs an HTML validation check before the accessibility checks, as valid HTML is a BBC requirement anyway.

**Comment by dd8** [comment]
>>If both the lang attribute in no namespace and the lang attribute in the XML namespace are set on an element, user agents must use the lang attribute in the XML namespace, and the lang attribute in no namespace must be ignored for the purposes of determining the element’s language.

> So xml:lang takes priority over lang, without any doubt.

It only takes priority if the document is processed as XHTML, and `xml:lang` is interpreted as the lang attribute in the XML namespace.

The `xml:lang` attribute has 2 different interpretations depending on whether the document is processed as HTML or as XHTML.

When served with `Content-Type: text\html` the document is processed as HTML and `xml:lang` is interpreted as the literal localname "xml:lang" in no namespace.
https://stackoverflow.com/questions/4320161/html5-html-attributes-xmlns-lang-xmllang/4320267#4320267

When served as `Content-Type: application/xhtml+xml`  the document is processed as XHTML and `xml:lang` is interpreted as the literal localname "lang" in the XML namespace. 
https://html.spec.whatwg.org/multipage/xhtml.html#the-xhtml-syntax

That's why the spec makes a distinction between these 2 cases, and has the note:
> The attribute in no namespace with no prefix and with the literal localname "xml:lang" has no effect on language processing.

HTML5 documents served as application/xhtml+xml are rare in practice because there are restrictions on scripting - you can't use `document.write()` and the `<noscript>` element.

**Comment by dd8** [comment]
The wording in the spec is very unclear, so these test files might make the issue a bit clearer.

https://qa.powermapper.com/Tests/ACT-R/xml-lang.html
https://qa.powermapper.com/Tests/ACT-R/xml-lang.xhtml

Both files are identical - and contain xml:lang='en' on the html element. The only difference is the file extension which controls the Content-Type returned for the file

https://qa.powermapper.com/Tests/ACT-R/xml-lang.html 
returns `Content-Type: text/html`
The language calculation returns null when the file is served as `text/html`, because `xml:lang` is in the null namespace, and the W3 validator returns a warning about missing a missing `lang` attribute.

https://qa.powermapper.com/Tests/ACT-R/xml-lang.xhtml 
returns `Content-Type: application/xhtml+xml`
The language calculation returns `en` when the file is served as `application/xhtml+xml`, because `xml:lang` is in the XML namespace, and the W3 validator returns no warnings.

**tl;dr** `xml:lang` is ignored for all pages served as `text/html` (which is nearly all pages on the web)

**Comment by EmmaJP** [comment]
So... is this issue less to do with what the rule is checking (in the case of https://act-rules.github.io/rules/bf051a that the value of the lang and/or xml:lang on the root html element of a page is valid) and more to do with the wording of the Accessibility Support section within the rule (which seems to be around whether or not and how some assistive technology supports the lang and/or xml:lang)?

Just asking, because to me whether or not the lang and/or xml:lang is or is not supported by a browser/AT combination seems to have very little bearing on whether or not the value of those attributes is valid. If a browser/AT combination does not support the lang and/or xml:lang attributes they will be ignored. However, if they are supported then it matters that the values are valid and meaningful to the tech that can do something with that information.

The WCAG guideline is 3.1.1 - the language of the page can be programmatically determined. Surely what happens beyond that should not be the concern of an ACT Rule.

**Comment by dd8** [comment]
The initial report was this wording in the rule, which doesn't match what the normative text in the HTML Standard says, so it's causing confusion:

>>While HTML5 specification indicates that xml:lang attribute takes priority over lang attribute, certain assistive technologies prioritise lang over xml:lang instead.

> whether or not the lang and/or xml:lang is or is not supported by a browser/AT combination seems to have very little bearing on whether or not the value of those attributes is valid

The comments are about being valid according to specification, not about accessibility support. 

>The WCAG guideline is 3.1.1 - the language of the page can be programmatically determined. >Surely what happens beyond that should not be the concern of an ACT Rule.

If programmatically determined follows the wording in the HTML Standard, then here's how language is determined by the standard:

- for pages served as `Content-Type: text/html` - the **`lang`** attribute is used and `xml:lang` is ignored for determining language
- for pages served as `Content-Type: application/xhtml+xml ` - the **`xml:lang`** attribute is used and `lang` ignored for determining language
- if `lang` and `xml:lang` are both specified they must have he same value (doing this ensures pages can be served with any Content-Type with no difference in the language determined)

This raises an issue for evaluation of SC 3.1.1. - you need to know the HTTP `Content-Type` the document will be served as, unless both `lang` and `xml:lang` are specified.

**Comment by EmmaJP** [comment]
A composite rule that is working out if 3.1.1 is being met or not may need to know the HTTP content-type ... but the three atomic rules do not. They only need to know if a lang attribute exists, if the value of the lang/xml:lang is valid, and if both exist that they have the same value.

**Comment by dd8** [comment]
There's a related problem in https://act-rules.github.io/rules/b5c3f8 - it passes if either `lang` or `xml:lang` are specified. The Assumption section should probably say that the attribute used should be the appropriate one for the Content-Type (if it doesn't match the attribute is ignored according to spec).

**Comment by WilcoFiers** [comment]
@dd8 I think you're right. I think there's actually a problem with this rule. Passed example 5:

```
<html lang="" xml:lang="nl"></html>
```

This should fail, because this isn't served as an XHTML page. I think the way to fix this is to put in a second expectation:

> If the [content type] is "text/html", the `lang` attribute must be set.

We'll have to figure out how to create `application/xml` test cases.

**Comment by dd8** [comment]
There might be some wider changes required - the references to `xml:lang` are being removed from H57 and H58:
https://github.com/w3c/wcag/issues/759
https://github.com/w3c/wcag/pull/766

There's still an open question on what happens with XML content inside HTML (i.e. SVG / MathML).

**Comment by dd8** [comment]
The issue is even more complicated once an HTML page contains `<svg>` elements

For HTML elements, *either* `lang` or `xml:lang` are used depending on Content-Type
For inline SVG elements embedded on HTML pages *both* `lang` are used `xml:lang`, with `xml:lang` taking precedence.

I've updated these test cases to show the differences
https://qa.powermapper.com/Tests/ACT-R/xml-lang.html
https://qa.powermapper.com/Tests/ACT-R/xml-lang.xhtml

The differences in interpretation are due to multiple W3 Recommendations covering use of `lang` vs `xml:lang` and none of them agreeing.

## XHTML 1.0
>Use both the lang and xml:lang attributes when specifying the language of an element. The value of the xml:lang attribute takes precedence.
https://www.w3.org/TR/xhtml1/#C_7

The XML 1.0 DTD has this definition allowing both lang and xml:lang:
 
<!ENTITY % i18n
 "lang        %LanguageCode; #IMPLIED
  xml:lang    %LanguageCode; #IMPLIED
  dir         (ltr|rtl)      #IMPLIED"
  >

## XHTML 1.1 First Edition
>On every element, the lang attribute has been removed in favor of the xml:lang attribute
https://www.w3.org/TR/2001/REC-xhtml11-20010531/changes.html#a_changes

## XHTML 1.1 Second Edition
The statement about removal of lang was deleted in Second Edition,
https://www.w3.org/TR/xhtml11/changes.html#a_changes

This definition was added back to the DTD allowing both lang and xml:lang

<!ENTITY % lang.attrib
     "xml:lang     %LanguageCode.datatype;  #IMPLIED
      lang         %LanguageCode.datatype;  #IMPLIED"
>

## HTML 5.0
Browsers should use ‘lang’ when document is served as ‘text/html’
Browsers should use ‘xml:lang’ when document is served as ‘application/xhtml+xml’
There’s a clear description here:
https://www.w3.org/International/questions/qa-html-language-declarations#attributes

**Note:** this doesn’t cover the case of SVG elements embedded in HTML (see below)

## SVG 1.1
Allows xml:lang only 
https://www.w3.org/TR/SVG11/struct.html#LangSpaceAttrs

## SVG 2.0
Allows both lang and xml:lang - xml:lang is used if both are specified
>If both the ‘lang’ attribute in no namespace and the ‘lang’ attribute in the XML namespace are set on an element, user agents must use the ‘lang’ attribute in the XML namespace, and the ‘lang’ attribute in no namespace must be ignored for the purposes of determining the element's language.
https://www.w3.org/TR/SVG2/struct.html#LangSpaceAttrs

**Note:** The W3 HTML5 validator report ‘lang’ on SVG elements as an error - see https://github.com/validator/validator/issues/848

**Comment by kasperisager** [comment]
This is... a mess. My recommendation would be to keep things simple and follow the requirements outlined by the HTML specification (read: https://html.spec.whatwg.org), as mentioned by @Jym77 :

> The `lang` attribute in the XML namespace may be used on HTML elements in XML documents, as well as elements in other namespaces if the relevant specifications allow it (in particular, MathML and SVG allow `lang` attributes in the XML namespace to be specified on their elements). If both the `lang` attribute in no namespace and the `lang` attribute in the XML namespace are specified on the same element, they must have exactly the same value when compared in an ASCII case-insensitive manner.
>
> Authors must not use the `lang` attribute in the XML namespace on HTML elements in HTML documents. To ease migration to and from XML, authors may specify an attribute in no namespace with no prefix and with the literal localname "`xml:lang`" on HTML elements in HTML documents, but such attributes must only be specified if a `lang` attribute in no namespace is also specified, and both attributes must have the same value when compared in an ASCII case-insensitive manner.

If so, the rule should only look at the `lang` attribute and ignore the `xml:lang` attribute. The rule https://act-rules.github.io/rules/5b7ae0 would then check that the `xml:lang` attribute, if present, matches the `lang` attribute.

**Comment by dd8** [comment]
https://act-rules.github.io/rules/5b7ae0/

Lang subtags are important for some languages

> Although there is no known way that an inappropriate secondary subtag in the lang or xml:lang attribute (such as "en-XYZ") can lead to accessibility issues, it is conceivable that with specific assistive technologies on some languages, there can be exceptions. In that case lang and xml:lang should have matching secondary subtags.

There are some languages where the subtag has an effect on how easily voicing is understood (e.g. `pt-BR` vs `pt-PT`)
https://travel.stackexchange.com/questions/13683/will-i-be-understood-in-all-portuguese-speaking-countries-if-i-learn-portuguese

There's a more extreme problem with  Mandarin ('zh-Cmn`) and Cantonese (`zh-Yue`) since these use the same *written* language but are spoken very differently (so different that Mandarin and Cantonese speakers cannot understand each other when talking)
https://www.quora.com/Do-Cantonese-speakers-understand-Mandarin

There's also the case of script which can form part of language tags (e.g. `en-Brai-US` for US english written in Braille)

**Comment by dd8** [comment]
https://act-rules.github.io/rules/5b7ae0/

Open question: does this rule affect accessibility? If user-agents behave as specified and nearly all content is served as `text\html` then `xml:lang` is nearly always ignored. 

This MDN page says that `xml:lang` is deprecated, but doesn't have a normative reference:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes

**Comment by dd8** [comment]
Here is Common Crawl data to back up the assumption that 'application/xhtml+xml' pages are rare: 

| Content-Type  | % of pages |
| ------------- | ------------- |
| text/html  | 97.61%  |
| application/xhtml+xml  | 0.038%  |
| application/xml  | 0.031%  |
| text/xml  | 0.084%  |

https://commoncrawl.github.io/cc-crawl-statistics/plots/mimetypes

**Comment by dd8** [comment]
For accessibility support I've looked at some browser source code.

There are 4 different language calculation algorithms in Firefox and Chrome which all produce different results:

- Firefox CSS : lang()
- Firefox Accessibility API language()
- Chrome CSS : lang()
- Chrome Accessibility API language()

# Firefox

Appears to calculate a different language for the CSS :lang() selector and the language exposed in the a11y tree. Needs confirmed with browser testing

## CSS :lang()
- uses `lang` in XML namespace (only for `application/xhtml+xml`), then
- uses `lang` in no namespace, then

For CSS :lang() it uses` nsAtom* nsIContent::GetLang()`
https://dxr.mozilla.org/mozilla-central/source/dom/base/FragmentOrElement.cpp#316

## Accessibility API Language
- only uses `lang` in no namespace

This means you can get different languages used in CSS and A11y for pages served as `application/xhtml+xml` 

# Chromium

Also seems to calculate a different language for the CSS :lang() selector and the language exposed in the a11y tree. Needs confirmed with browser testing

## CSS :lang()
- uses `lang` in XML namespace (only for `application/xhtml+xml`), then
- uses `lang` in no namespace, then
- uses uses meta http-equiv=Content-Language element, then

`Element::ComputeInheritedLanguage() `
https://github.com/chromium/chromium/blob/bed82fec55aabe34b78f894068e3ce1dcc67ac6a/third_party/blink/renderer/core/dom/element.cc#L4817

## Accessibility API Language
- uses `lang` in no namespace, then
- uses uses meta http-equiv=Content-Language element, then
- first language in HTTP Accept-Languages header, then
- browser UI language

`AXObject::Language()`
https://github.com/chromium/chromium/blob/92990c67250df1e6c4bdc70de41e288e47e3c636/third_party/blink/renderer/modules/accessibility/ax_object.cc#L2575

**Comment by Jym77** [comment]
Isn't this _de facto_ fixed by #934 (which removes mentions of `xml:lang` and thus makes the whole problem disappear…)?

**Comment by dd8** [comment]
Yes, once https://github.com/act-rules/act-rules.github.io/pull/934 is merged this can be closed

**Comment by jkodu** [comment]
PR is in progress, closing issue.
